### PR TITLE
[SPARK-21893][BUILD][STREAMING][WIP] Put Kafka 0.8 behind a profile

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

Put Kafka 0.8 support behind a kafka-0-8 profile.

## How was this patch tested?

Existing tests, but, until PR builder and Jenkins configs are updated the effect here is to not build or test Kafka 0.8 support at all.

### CODE REVIEW

File: dev/create-release/release-build.sh [hunk 1]
```diff
@@ -80,8 +80,17 @@ NEXUS_PROFILE=d63f592e7eac0 # Profile for Spark staging uploads
 BASE_DIR=$(pwd)
 
 MVN="build/mvn --force"
-PUBLISH_PROFILES="-Pmesos -Pyarn -Phive -Phive-thriftserver"
-PUBLISH_PROFILES="$PUBLISH_PROFILES -Pspark-ganglia-lgpl -Pkinesis-asl"
+
```
Comment (srowen) [inline dev/create-release/release-build.sh]: There should be no actual change to the profiles that are set, except that the current Scala 2.11 build adds `-Pkafka-0-8` I rearranged these to make it easier to reason about the profiles in one place.

**Review by srowen — COMMENTED** [review]
Note that this change must be accompanied by adding -Pkafka-0-8 to all Jenkins jobs on master, and PR builders.

**Comment by jerryshao** [comment]
@srowen how do you handle python kafka.py, should it also be opt-in? As far as I understand looks like you don't address it in this PR.

**Comment by srowen** [comment]
Ah, this is where I need some advice. Is there an element of the Pyspark code that has to vary too? obviously Python itself isn't dependent on Scala, but the tests obviously can't test 0.8 if it's not there.  

Yes I think there are a few more issues to iron out before this works, so input is welcome.

**Comment by koeninger** [comment]
I think it makes sense to go ahead and put deprecation warnings in this PR as well

**Comment by vanzin** [comment]
For python I think you just do what kinesis does:

```
# Must be same as the variable and condition defined in KinesisTestUtils.scala
kinesis_test_environ_var = "ENABLE_KINESIS_TESTS"
are_kinesis_tests_enabled = os.environ.get(kinesis_test_environ_var) == '1'
```

The comment is misleading though, since that variable is set in `modules.py` which you are already modifying.

**Comment by felixcheung** [comment]
re kafka -https://github.com/apache/spark/pull/19134#issuecomment-327180847 - pretty sure that is still needed for DStream - just as any API we shouldn't just remove it in point releases...

**Comment by jerryshao** [comment]
I see @felixcheung . Since we have a solution to turn off Python Kafka unit test as mentioned by @vanzin , so it is fine to just mark as deprecated, not remove the code.

Another thing is that we also have some Kafka related examples, maybe we should also handle them.

**Comment by srowen** [comment]
Good feedback, all. I think I can fix the MiMa failure, and now the Python tests. Let me try to get that far.

Yes, the examples use Kafka 0.8 in some places. It seems excessive to set up a new optional module of Kafka 0.8-only examples, so, perhaps it's time to remove them?

Because there's also the suggestion to deprecate 0.8. And then that entails making 0.10 stable.

File: python/pyspark/streaming/tests.py [hunk 2]
```diff
@@ -1051,6 +1062,9 @@ def test_kafka_stream(self):
 
     def test_kafka_direct_stream(self):
         """Test the Python direct Kafka stream API."""
+        if not are_kafka_tests_enabled:
```
Comment (vanzin) [inline python/pyspark/streaming/tests.py]: Wouldn't it be simpler to follow the kinesis example?

The code checks this once in the `__main__` section, and if tests are disabled the suite is not added to the list of suites to be run.

Also, am I reading things wrong, or there are no tests for the 0.10 backend here? Is the 0.10 backend supported in pyspark?

File: python/pyspark/streaming/tests.py [hunk 3]
```diff
@@ -1003,6 +1003,11 @@ def check_output(n):
         self.ssc.stop(True, True)
 
 
+# Must be same as the variable and condition defined in modules.py
+kafka_test_environ_var = "ENABLE_KAFKA_TESTS"
```
Comment (vanzin) [inline python/pyspark/streaming/tests.py]: I'd call this `ENABLE_KAFKA_0_8_TESTS`, ignoring my question about the existence of 0.10 tests in pyspark.

File: python/pyspark/streaming/tests.py (hunk 2 above)
Comment (srowen) [reply python/pyspark/streaming/tests.py]: @vanzin ah I was distracted by how it's also handled in KinesisStreamTests. Done, I think.

I think that yes there is no 0.10 test, as I don't think Pyspark supports 0.10 (right @jerryshao ?): https://github.com/apache/spark/pull/14340#issuecomment-235457517

So all Kafka Pyspark tests should be disabled, yeah.
But, does this mean that Pyspark Kafka support is effectively deprecated, entirely?

(I need to add deprecation messages to 0.8 code)

**Comment by vanzin** [comment]
LGTM. The pyspark vs. 0.10 discussion seems like it belongs in a separate bug.

**Comment by koeninger** [comment]
There's already a jira about why 0.10 doesn't have python support, https://issues-test.apache.org/jira/browse/SPARK-16534

**Comment by srowen** [comment]
OK by me, if it's all on-purpose. So, am I right that all of the Pyspark Kafka integration is effectively deprecated as of this change, because it depends on now-deprecated 0.8 support? just want to make sure I'm reading that right, and if so I'll try to mark it as such.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**